DPoP Optimisation (#146)
oauth-wg/draft-ietf-oauth-attestation-based-client-auth: draft-ietf-oauth-attestation-based-client-auth.md

(5) The Client Instance generates a Proof of Possession (PoP) with the Client Instance Key.

> (6) The Client Instance sends the Client Attestation JWT along with its Proof of Possession to the Authorization Server, e.g. within a token request. The Proof of Possession is either a Client Attestation PoP JWT or a DPoP proof when operating in combined mode (see [](#dpop-combined-mode)). The Authorization Server validates the Client Attestation and thus authenticates the Client Instance.

Reviewer: ```suggestion
(6) The Client Instance sends the Client Attestation JWT along with its Proof of Possession to the Authorization Server, e.g. within a token request. The Proof of Possession is either a Client Attestation PoP JWT or a DPoP proof. The Authorization Server validates the Client Attestation and thus authenticates the Client Instance.
```

                                     "_" / "~" / "+" / "/" ) *"="
~~~

## Using DPoP as the Proof of Possession {#dpop-combined-mode}

> This section defines an optimization that allows a single Proof of Possession (PoP) JWT to satisfy the role of both (a) the Client Attestation PoP defined in this specification and (b) the DPoP proof defined in {{RFC9449}}. In this "combined mode" the Client Instance Key and the DPoP Key are the same asymmetric key pair, and a request using the mechanism carries only one PoP, the DPoP proof, instead of two separate PoP JWTs (the DPoP proof and Client Attestation PoP).

Reviewer: ```suggestion
This section defines an optimization that allows a single Proof of Possession (PoP) JWT to satisfy the role of both (a) the Client Attestation PoP defined in this specification and (b) the DPoP proof defined in {{RFC9449}} for sender-contrained access tokens. In this "combined mode" the Client Instance Key and the DPoP Key are the same asymmetric key pair, and a request using the mechanism carries only one PoP, the DPoP proof, instead of two separate PoP JWTs (the DPoP proof and Client Attestation PoP JWT).
```

The following rules apply to the DPoP proof as defined in {{RFC9449}}:

1. The DPoP proof MUST adhere to {{RFC9449}}
2. The public key located in the DPoP proof MUST match the public key located in the `cnf` claim of the Client Attestation JWT.

The following non-normative example shows a token request using combined mode (line breaks for display only):

~~~
POST /token HTTP/1.1
Host: as.example.com
Content-Type: application/x-www-form-urlencoded
OAuth-Client-Attestation: <Client-Attestation-JWT>

> DPoP: <Combined-DPoP-And-Attestation-PoP-JWT>

Reviewer: Let's fill these with actual examples

&code_challenge_method=S256&scope=account-information
~~~

## Usage of Client Instance Key as DPoP Key {#combined-dpop}

This section defines an OPTIONAL optimization that allows a single Proof of Possession (PoP) JWT to satisfy both (a) the Client Attestation PoP defined in this specification and (b) the DPoP proof defined in {{RFC9449}}. In this "combined mode" the Client Instance Key and the DPoP Key are the same asymmetric key pair, and a request using the mechanism carries only one PoP, the DPoP proof, instead of two separate PoP JWTs (the DPoP proof and Client Attestation PoP).

### Validating HTTP requests

In order to detect whether this combined mode is being used, the following conditions MUST be met:
1. There is precisely one OAuth-Client-Attestation HTTP request header field, where its value is a single well-formed JWT conforming to the syntax outlined in [](#client-attestation-jwt).
2. There is precisely one DPoP HTTP request header field, its value is conforming to the syntax outlined in {{RFC9449}}.
3. There is no OAuth-Client-Attestation-PoP HTTP request header field present.

Provided the above conditions are meet the following additional rules apply:

> 1. The public key located in the DPoP proof MUST match the public key located in the `cnf` claim of the Client Attestation JWT.

Author: Should we be more specific around how this matching should occur, e.g if both JWK use thumbprint etc?

1. The audience claim in the Client Attestation PoP JWT is the issuer identifier URL of the Authorization Server as described in {{RFC8414}}.

> 1. Depending on the security requirements of the deployment, additional checks to guarantee replay protection for the Client Attestation PoP JWT might need to be applied (see [](#security-consideration-replay) for more details).

Reviewer: ```suggestion
1. If the Client fetched a challenge in accordance with [](#challenge-retrieval) or received a challenge in accordance with [](#challenge-header), it MUST match the challenge claim of the Client Attestation PoP JWT.
1. Depending on the security requirements of the deployment, additional checks to guarantee replay protection for the Client Attestation PoP JWT might need to be applied (see [](#security-consideration-replay) for more details).
```

## DPoP Combined Mode {#verification-dpop-combined}

This section applies when the DPoP combined mode is used as defined in [](#dpop-combined-mode). When the Client Attestation PoP JWT is used as the Proof of Possession instead, this section does not apply; see [](#verification-client-attestation-pop-jwt).

To validate a request using DPoP combined mode, the receiving server MUST perform the following steps:

> 1. Validate the Client Attestation JWT as specified in [](#verification-client-attestation-jwt).

Reviewer: redundant
```suggestion
```

1. There is no `OAuth-Client-Attestation-PoP` HTTP request header field present in the request.
1. There is precisely one `DPoP` HTTP request header field present in the request.
1. Validate the DPoP proof in accordance with {{RFC9449}}.

> 1. The public key in the `jwk` header parameter of the DPoP proof MUST be identical to the public key in the `cnf` claim of the Client Attestation JWT.

Reviewer: ```suggestion
1. The public key in the `jwk` header parameter of the DPoP proof MUST be identical to the public key in the `cnf` claim of the Client Attestation JWT.
1. If the Client fetched a challenge in accordance with [](#challenge-retrieval) or received a challenge in accordance with [](#challenge-header), it MUST match the nonce payload claim of the DPoP proof.
```

# Authorization Server Metadata {#as-metadata}

The Authorization Server SHOULD communicate support for authentication with Attestation-Based Client Authentication using a Client Attestation PoP JWT as the PoP by using the value `attest_jwt_client_auth` in the `token_endpoint_auth_methods_supported` within its published metadata. The Authorization Server SHOULD communicate support for authentication with Attestation-Based Client Authentication using a DPoP proof as the PoP by using the value `attest_jwt_client_auth_dpop` in the `token_endpoint_auth_methods_supported` within its published metadata. The client SHOULD fetch and parse the Authorization Server metadata and recognize Attestation-Based Client Authentication as a client authentication mechanism if either of the given `token_endpoint_auth_methods_supported` values are present.

> The Authorization Server SHOULD communicate supported algorithms for client attestations by using `client_attestation_signing_alg_values_supported` and `client_attestation_pop_signing_alg_values_supported` within its published metadata. This enables the client to validate that its client attestation is understood by the Authorization Server prior to authentication. The client MAY try to get a new client attestation with different algorithms. The Authorization Server MUST include `client_attestation_signing_alg_values_supported` and `client_attestation_pop_signing_alg_values_supported` in its published metadata if the `token_endpoint_auth_methods_supported` includes `attest_jwt_client_auth`. When `token_endpoint_auth_methods_supported` includes `attest_jwt_client_auth_dpop`, the Authorization Server SHOULD advertise supported DPoP signing algorithms using `dpop_signing_alg_values_supported` as defined in {{RFC9449}}, rather than `client_attestation_pop_signing_alg_values_supported`, as the Proof of Possession in combined mode is a DPoP proof.

Reviewer: Just to mention that there is proposal to simplify: https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/issues/170

# Implementation Considerations

## DPoP Combined Mode Considerations

When using DPoP combined mode, the key used for client authentication and token binding is shared. This may be undesirable depending on the deployment considerations of the Client. Conversely, the benefits of this approach are as follows:

* It authenticates (attests) the DPoP key used for sender-constraining tokens against the Client deployment.

> * It can reduce implementation complexity for the Client by reducing the number of JWTs that need to be constructed or validated in a request.

Reviewer: ```suggestion
* It reduces implementation complexity for the Client by minimizing the number of JWTs that need to be constructed or validated in a request.
* It reduces run-time costs for the Client by minimizing the number of cryptographic operations that need to be constructed in a request, especially if the keys are in a remote and/or hardware-backed key storage.
```

Upon receiving a Client Attestation, the receiving server MUST ensure the following conditions and rules:

1. If the Client Attestation was received via Header based Syntax (as described in [](#header-based)) the HTTP request MUST conform to exactly one of the following:

  * Separate proofs pattern: exactly one field `OAuth-Client-Attestation` and exactly one field `OAuth-Client-Attestation-PoP`, each containing exactly one well-formed JWT; OR

>   * Combined DPoP pattern: exactly one field `OAuth-Client-Attestation` and exactly one field `DPoP` meeting the requirements of [](#combined-dpop), and no `OAuth-Client-Attestation-PoP` field present.

Author: Note im definitely not convinced of the changes here, it implies the combined mode isn't really optional
